src/Lean/Parser/Module.lean

```diff
@@ -103,6 +109,9 @@ partial def parseCommand (inputCtx : InputContext) (pmctx : ParserModuleContext)
     let pos' := pos
     let p := andthenFn whitespace topLevelCommandParserFn
     let s := p.run inputCtx pmctx (getTokenTable pmctx.env) { cache := initCacheForInput inputCtx.input, pos }
+    -- save errors from sub-recoveries
+    for (rpos, rstk, recovered) in s.recoveredErrors do
+      messages := messages.add <| mkErrorMessage inputCtx rpos rstk recovered
```

Comment:
Should this be guarded by `unless recovering && ignore` as well?

Reply:
I don't think so. The idea here is that if any errors were recovered from while parsing the command, they need to end up in the message log. But `recovering` says that the parser had failed to parse the command, and it's attempting to resynchronize. The guarded case below is there to ignore errors that occur as a part of the outer loop's recovery process, but this code is intended to save errors that occurred as part of the "ordinary" us of the parser.

Does that make sense?

Reply:
I'm not sure I follow, we don't know whether we're in the "ordinary" case unless we use that specific conditional. But it seems unlikely that `getPos?` would be empty if we did recover from errors during the parse so in practice it should be okay.

Reply:
I mean, passing the tests and seeming to work is necessary, but not sufficient. There could totally be an obscure bug if we don't understand why this is where it is, so it's worth hashing out.

The way it's written right now, every invocation of the top-level command parser may end up in the following states:
1. No errors at all - then the `for` executes 0 times, no errors to report. ✅ 
2. A successful parse, but errors were recovered from - these belong in the message log, and the elaborator will bump into some `missing`s, but we don't need to go into recovery mode at the top level.
3. An unsuccessful parse, whether or not errors were recovered - in this case the recovered errors need reporting, and the outer loop needs to go into recovery mode

I think this is what's going on here. But it sounds like we're talking past each other, or one or both of us has misunderstood something. Can you say more about why it should perhaps be guarded?

Reply:
In the end, we recover from every parse error, so "recovered errors" is a slightly confusing description to me - I would expect either all or none of `s.recoveredErrors` and `s.errorMsg` to be reported for a single command parser invocation, which is not the case in your current code. But as I said, I can't think of a practical issue here - it is hard to imagine that we would ever want to use `recover` without already having consumed at least one token, which should lead to `ignore` being false. So I don't have a clear recommendation except "looked potentially inconsistent with existing code to me at first glance".